### PR TITLE
wip: neato

### PR DESCRIPTION
**Pull request by rjrjr**

### CODE REVIEW

File: kotlin/workflow-runtime/src/main/java/com/squareup/workflow/LaunchWorkflow.kt [hunk 1]
```diff
@@ -140,11 +140,11 @@ fun <InputT, OutputT : Any, RenderingT, ResultT> launchSingleOutputWorkflowIn(
     output: Deferred<OutputT>
   ) -> ResultT
 ): ResultT = launchWorkflowIn(scope, workflow, inputs, initialSnapshot) { renderings, outputs ->
-  val workflowScope = this
+//  val workflowScope = this
```
Comment (rjrjr) [inline kotlin/workflow-runtime/src/main/java/com/squareup/workflow/LaunchWorkflow.kt]: @zach-klippenstein We're nearly there, but there's a teardown race or something going on. If I don't comment out this block, the new `output.await()` call in `WorfklowRunnerViewModel` never fires, and so the call to `finish()` in `HelloWorkflowActivity` is never made, and the app hangs.

Comment (rjrjr) [reply kotlin/workflow-runtime/src/main/java/com/squareup/workflow/LaunchWorkflow.kt]: Maybe we should just implement this behavior in the view model.

Comment (rjrjr) [reply kotlin/workflow-runtime/src/main/java/com/squareup/workflow/LaunchWorkflow.kt]: Yeah, that works nicely. I think we'll want to revert the `launchWorkflowIn` PR, just lost our only use case.

Comment (zach-klippenstein) [reply kotlin/workflow-runtime/src/main/java/com/squareup/workflow/LaunchWorkflow.kt]: Did I miss a test case?

File: kotlin/workflow-ui-android/src/main/java/com/squareup/workflow/ui/WorkflowRunnerViewModel.kt [hunk 2]
```diff
@@ -74,16 +72,27 @@ internal class WorkflowRunnerViewModel<OutputT : Any>(
...
-      WorkflowRunnerViewModel(viewRegistry, renderings, outputs, this) as T
+      WorkflowRunnerViewModel(renderings, output, this) as T
     }
   }
 
-  private val snapshotJob = scope.launch {
-    renderingsFlow
-        .map { it.snapshot }
-        .collect { lastSnapshot = it }
+  private val _result = BehaviorSubject.create<OutputT>()
```
Comment (zach-klippenstein) [inline kotlin/workflow-ui-android/src/main/java/com/squareup/workflow/ui/WorkflowRunnerViewModel.kt]: Use `SingleSubject`

File: kotlin/workflow-runtime/src/main/java/com/squareup/workflow/LaunchWorkflow.kt (hunk 1 above)
Comment (zach-klippenstein) [reply kotlin/workflow-runtime/src/main/java/com/squareup/workflow/LaunchWorkflow.kt]: Ok, this doesn't actually surprise me. The cancellation does feel pretty racy. It's easy enough to do this manually that providing an API just to do it is redundant and over-engineered anyway.

File: kotlin/workflow-ui-android/src/main/java/com/squareup/workflow/ui/WorkflowRunnerViewModel.kt (hunk 2 above)
Comment (zach-klippenstein) [reply kotlin/workflow-ui-android/src/main/java/com/squareup/workflow/ui/WorkflowRunnerViewModel.kt]: Or better yet, `scope.rxSingle`

File: kotlin/workflow-ui-android/src/main/java/com/squareup/workflow/ui/WorkflowRunnerViewModel.kt [hunk 3]
```diff
@@ -74,16 +72,27 @@ internal class WorkflowRunnerViewModel<OutputT : Any>(
...
+  override val result: Single<out OutputT> = _result.firstOrError()
+
+  init {
+    scope.launch {
+      renderingsFlow
+          .map { it.snapshot }
+          .collect { lastSnapshot = it }
+    }
```
Comment (zach-klippenstein) [inline kotlin/workflow-ui-android/src/main/java/com/squareup/workflow/ui/WorkflowRunnerViewModel.kt]: Slightly more idiomatic/less nesting:
```kotlin
renderingsFlow.map { it.snapshot }
  .onEach { lastSnapshot = it }
  .launchIn(scope)
```

File: kotlin/workflow-ui-android/src/main/java/com/squareup/workflow/ui/WorkflowRunnerViewModel.kt (hunk 2 above)
Comment (zach-klippenstein) [reply kotlin/workflow-ui-android/src/main/java/com/squareup/workflow/ui/WorkflowRunnerViewModel.kt]: Btw looks like these builders are about to lose their scope receiver:

https://github.com/Kotlin/kotlinx.coroutines/commits/reactive-scopeless

File: kotlin/workflow-runtime/src/main/java/com/squareup/workflow/LaunchWorkflow.kt (hunk 1 above)
Comment (rjrjr) [reply kotlin/workflow-runtime/src/main/java/com/squareup/workflow/LaunchWorkflow.kt]: https://github.com/square/workflow/pull/462

**Pull request closed without merging**